### DISCUSSION TITLE
Confused about pytorch examples

### DISCUSSION THREAD

**Discussion by chdamianos** [horovod/horovod#2868] [Q&A]
I've looked at the PyTorch examples at 
https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_imagenet_resnet50.py 
https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_mnist.py 
https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_synthetic_benchmark.py
and I'm confused about how to distribute the optimizer:

Specifically for the pytorch_imagenet_resnet50.py example at https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_imagenet_resnet50.py#L264-L290 
1. first the Pytorch optimizer is defined
```python
optimizer = optim.SGD(model.parameters(),
                      lr=(args.base_lr *
                          lr_scaler),
                      momentum=args.momentum, weight_decay=args.wd)
```
2. the optmizer is then reassigned using hvd.DistributedOptimizer 
```python
optimizer = hvd.DistributedOptimizer(
    optimizer, named_parameters=model.named_parameters(),
    compression=compression,
    backward_passes_per_step=args.batches_per_allreduce,
    op=hvd.Adasum if args.use_adasum else hvd.Average,
    gradient_predivide_factor=args.gradient_predivide_factor)
```
3. and then hvd.broadcast_optimizer_state is called. 
```python
hvd.broadcast_optimizer_state(optimizer, root_rank=0)
```
Same order for the pytorch_synthetic_benchmark.py example

In the pytorch_mnist.py example the order is different https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_mnist.py#L178-L189 
1. first the Pytorch optimizer is defined
```python
optimizer = optim.SGD(model.parameters(), lr=args.lr * lr_scaler,
                      momentum=args.momentum)
```
2. then hvd.broadcast_optimizer_state is called. 
```python
hvd.broadcast_optimizer_state(optimizer, root_rank=0)
```
3. the optimizer is then reassigned using hvd.DistributedOptimizer 
```python
optimizer = hvd.DistributedOptimizer(optimizer,
                                     named_parameters=model.named_parameters(),
                                     compression=compression,
                                     op=hvd.Adasum if args.use_adasum else hvd.Average,
                                     gradient_predivide_factor=args.gradient_predivide_factor)
```

Which order is the correct one?
*  Define PyTorch optimizer, Reassign using  hvd.DistributedOptimizer and broadcast using hvd.broadcast_optimizer_state
or
*  Define PyTorch optimizer, broadcast using hvd.broadcast_optimizer_state and then reassign using hvd.DistributedOptimizer

**Comment by maxhgerlach** [comment]
From reading the source I'd say either order is fine: https://github.com/horovod/horovod/blob/386be429b1417a1f6cb5e715bbe36efd2e74f402/horovod/torch/functions.py#L89

**Reply by chdamianos** [reply]
cool thanks @maxhgerlach for the reply!
